File: backend/semgrep_rules/tests/java/kisa_sw_05_os_command_injection.java
import javax.servlet.http.HttpServletRequest;

class kisa_sw_05_os_command_injection {

    Process unsafeConcat(
        HttpServletRequest req
    ) throws Exception {
        String host =
            req.getParameter("host");

        // ruleid: kisa.sw05.java.external-input-to-command
        return Runtime.getRuntime().exec("ping " + host);
    }


    Process unsafeDirectCommand(
        HttpServletRequest req
    ) throws Exception {
        String command =
            req.getParameter("command");

        // ruleid: kisa.sw05.java.external-input-to-command
        return Runtime.getRuntime().exec(command);
    }


    Process unsafeIntermediate(
        HttpServletRequest req
    ) throws Exception {
        String host =
            req.getHeader("X-Host");

        String command =
            "nslookup " + host;

        String copiedCommand =
            command;

        // ruleid: kisa.sw05.java.external-input-to-command
        return Runtime.getRuntime().exec(copiedCommand);
    }


    Process unsafeDynamicExecutable(
        HttpServletRequest req
    ) throws Exception {
        String executable =
            req.getParameter("bin");

        // ruleid: kisa.sw05.java.external-input-to-command
        return new ProcessBuilder(executable, "--version").start();
    }


    Process unsafeShellCommand(
        HttpServletRequest req
    ) throws Exception {
        String host =
            req.getParameter("host");

        String command =
            "ping " + host;

        // ruleid: kisa.sw05.java.external-input-to-command
        return new ProcessBuilder("sh", "-c", command).start();
    }


    Process safeSeparatedArgument(
        HttpServletRequest req
    ) throws Exception {
        String host =
            req.getParameter("host");

        // ok: kisa.sw05.java.external-input-to-command
        return new ProcessBuilder(
            "ping",
            host
        ).start();
    }


    Process safeHardcodedRuntime()
        throws Exception {

        // ok: kisa.sw05.java.external-input-to-command
        return Runtime.getRuntime().exec("uptime");
    }


    Process safeHardcodedProcessBuilder()
        throws Exception {

        // ok: kisa.sw05.java.external-input-to-command
        return new ProcessBuilder(
            "ping",
            "127.0.0.1"
        ).start();
    }
}

File: backend/semgrep_rules/tests/javascript/kisa_sw_05_os_command_injection.js
const childProcess =
  require("child_process");

const {
  exec,
  execSync,
} = require("child_process");


function unsafeObjectExec(req) {
  const host =
    req.query.host;

  // ruleid: kisa.sw05.javascript.external-input-to-command
  return childProcess.exec("ping " + host);
}


function unsafeDestructuredExec(req) {
  const host =
    req.body.host;

  const command =
    `nslookup ${host}`;

  // ruleid: kisa.sw05.javascript.external-input-to-command
  return exec(command);
}


function unsafeMultiStep(req) {
  const host =
    req.params.host;

  const command =
    "ping " + host;

  const copiedCommand =
    command;

  // ruleid: kisa.sw05.javascript.external-input-to-command
  return execSync(copiedCommand);
}


function unsafeSpawnShellCommand(req) {
  const command =
    req.query.command;

  // ruleid: kisa.sw05.javascript.external-input-to-command
  return childProcess.spawn(command, [], { shell: true });
}


function unsafeSpawnShellArgument(req) {
  const host =
    req.query.host;

  const args = [
    "-c",
    "1",
    host,
  ];

  // ruleid: kisa.sw05.javascript.external-input-to-command
  return childProcess.spawn("ping", args, { shell: true });
}


function safeSpawnWithoutShell(req) {
  const host =
    req.query.host;

  // ok: kisa.sw05.javascript.external-input-to-command
  return childProcess.spawn(
    "ping",
    [
      "-c",
      "1",
      host,
    ],
    {
      shell: false,
    }
  );
}


function safeExecFile(req) {
  const host =
    req.query.host;

  // ok: kisa.sw05.javascript.external-input-to-command
  return childProcess.execFile(
    "ping",
    [
      "-c",
      "1",
      host,
    ]
  );
}


function safeHardcodedExec() {
  // ok: kisa.sw05.javascript.external-input-to-command
  return childProcess.exec("uptime");
}


function safeUnrelatedMethod(req) {
  const command =
    req.query.command;

  const runner = {
    exec(value) {
      return value;
    },
  };

  // ok: kisa.sw05.javascript.external-input-to-command
  return runner.exec(command);
}
